src/library/scala/concurrent/Future.scala

```diff
@@ -193,8 +186,9 @@ trait Future[+T] extends Awaitable[T] {
    */
   def transform[S](s: T => S, f: Throwable => Throwable)(implicit executor: ExecutionContext): Future[S] =
     transform {
-      case Success(r) => Try(s(r))
-      case Failure(t) => Try(throw f(t)) // will throw fatal errors!
+      t =>
+        if (t.isInstanceOf[Success[T]]) t map s
+        else throw f(t.asInstanceOf[Failure[T]].exception) // will throw fatal errors!
```

Comment:
I am imagining that all these `asInstanceOf`s and `isInstanceOf`s are going to cause unchecked warnings - will probably need to `import scala.{unchecked => uc}` and throw a bunch of `@uc`s around.

Reply:
I'm wondering why is `isInstanceOf` used? wouldn't it be cheaper to do a check like
```
if t.isSuccess() => t map s
else throw f (t.asInstanceOf[Failure[T]].exception)
```

unless of course the isSuccess method has been removed in 2.13

Reply:
@javax-swing @NthPortal The answer is performance :)

Reply:
(good to comment these little performance hacks so someone doesn't innocently regress them later)

Reply:
My guess is that `isInstanceOf` translates to a single `instanceof` bytecode, while `isSuccess` is a method call which (a) allocates a stack frame and (b) has more than a single bytecode of body. Someone with more knowledge can correct me though

Reply:
@viktorklang If you want, I can add the `@unchecked` annotations this weekend, and either force push or push a commit labeled "[squash]" to your branch

Reply:
@NthPortal AFAIK there are no unchecked warnings, can you confirm?

Reply:
@SethTisue Basically the entire file is a big performance "hack". I can add a disclaimer in the beginning of the file?

Reply:
@viktorklang you're right; I was mistaken. I thought `isInstanceOf[Link[T]]` would warn because `T` is erased, but I was mistaken.